src/Command/RunCommand.php

```diff
@@ -69,24 +71,63 @@ public function terminationHandler(int $signalNumber): void
         $this->logger->notice(sprintf('Received signal "%s"', $signalNumber));
         $this->logProcessor->setLogInfo(new LogInfo($this->jobId, '', ''));
         try {
-            $jobStatus = $this->queueClient->getJob($this->jobId)->getStatus();
-            if ($jobStatus !== JobInterface::STATUS_TERMINATING) {
-                $this->logger->info(
-                    sprintf('Job "%s" is in status "%s", letting the job to finish.', $this->jobId, $jobStatus),
-                );
-                return;
-            }
+            $job = $this->queueClient->getJob($this->jobId);
         } catch (ClientException $e) {
             $this->logger->error(sprintf('Failed to get job "%s" for cleanup: ' . $e->getMessage(), $this->jobId));
             // we don't want the handler to crash
             return;
         }
+
+        $jobStatus = $job->getStatus();
+        if ($jobStatus !== JobInterface::STATUS_TERMINATING) {
+            $this->logger->info(
+                sprintf('Job "%s" is in status "%s", letting the job to finish.', $this->jobId, $jobStatus),
+            );
+            return;
+        }
+
+        // set up logging to storage API
+        $this->logProcessor->setLogInfo(new LogInfo(
+            $job->getId(),
+            $job->getComponentId(),
+            $job->getProjectId(),
+        ));
+        $options = BuildBranchClientOptionsHelper::buildFromJob($job)->setToken($this->storageApiToken);
+        $loggerService = $this->getLoggerService($options);
+        $clientWrapper = $this->getClientWrapper($options);
+
+        // set up runner
+        $component = $this->getComponentClass($clientWrapper, $job);
+        $jobDefinitions = $this->jobDefinitionFactory->createFromJob($component, $job, $clientWrapper);
+
+        $runner = new Runner(
+            $this->objectEncryptor,
+            $clientWrapper,
+            $loggerService,
+            new OutputFilter(60000),
+            $this->instanceLimits,
+        );
+        foreach ($jobDefinitions as $jobDefinition) {
+            $runner->cleanup($jobDefinition);
+        }
+
         $this->logger->info(sprintf('Terminating containers for job "%s".', $this->jobId));
+        $containerIds = $this->getContainerIds();
+
+        foreach ($containerIds as $containerId) {
+            $this->terminateContainer($containerId);
+        }
+        $this->logger->info(sprintf('Finished container cleanup for job "%s".', $this->jobId));
```

Comment:
Co tohle zavrit dovnitr do `Runner` class? Predstavoval bych si to neco jako:
* na `RunCommand` pribude property `private ?Runner $runner`, ktera se bude inicializovat v `execute`, kdyz se ten `Runner` vytvori
* tady v `terminationHandler` by si mel jen `if ($this->runner !== null) $this->runner->cleanup()`
* `Runner` by si interne trackoval aktualne bezici `JobDefinition` (pousti je postupne), resp. `$outputs`
* kdyz zavolas `$runner->cleanup()`, obcahoval by neoc podobnyho, jako `waitForStorageJobs`, jen misto `wait` by ty storage joby terminoval a pak udelal `cleanWorkspace`

Reply:
storage joby terminovat nejdou :) 

no za mě problém je, že to zase běží asynchronně, takže nikdy nevíš ve kterým okamžiku na co to přesně šaháš - on bude třeba $runner vytvořenej, ale buhví co bude zrovna dělat (bude mít inicalizovaný jobDefinitions?) 

stejně by se to teda mělo přehodit - napřed terminovat kontejnery a pak ty workspacy

Reply:
No jede to asynchronne, ale ne paralelne.. takze cokoliv potrebujes si muzes poznamenant do properties v runneru (napr. aktualne bezici jobDefinition, nebo seznam uz spustenych jobDefinitions atp.) a kdyz se ti pusti cleanup, tak vis, ze je to plany a nic jinyho nez cleanup v tu chvili nebezi.